layers/cmd/pipeline.go

```diff
@@ -14,18 +15,68 @@ import (
 
 type PipelineContext struct {
 	*ls.Context
-	Graph       graph.Graph
-	Roots       []graph.Node
+	graph       graph.Graph
+	roots       []graph.Node
 	InputFiles  []string
 	currentStep int
 	steps       []Step
 	Properties  map[string]interface{}
+	mu          sync.RWMutex
 }
 
 type Step interface {
 	Run(*PipelineContext) error
 }
 
+type ForkStep struct {
+	Steps [][]Step
+}
+
+func (fork ForkStep) Run(ctx *PipelineContext) error {
+	var wg sync.WaitGroup
+	wg.Add(len(fork.Steps))
+	errs := make([]error, 0, len(fork.Steps))
+	for _, pipe := range fork.Steps {
+		go func(steps []Step, currCtx *PipelineContext) {
+			pctx := &PipelineContext{
+				Context:     getContext(),
+				InputFiles:  make([]string, 0),
+				currentStep: 0,
+				Properties:  make(map[string]interface{}),
+			}
+			switch currCtx.graph {
```

Comment:
What exactly is this doing?

You need to keep track of which pipeline context owns the graph, something like:
```
type PipelineContext struct {
    graphOwner *PipelineContext
}
```
Given ctx *PipelineContext, ctx.GetGraphRW should clone the graph if ctx!=ctx.graphOwner, and set the graphOwner to ctx. You need mutex during that operation to make sure no one else has write lock on it.